### DISCUSSION TITLE
Recommended solution for Node.js APM?

### DISCUSSION THREAD

**Discussion by aecorredor** [vercel/vercel#4147] [General]
Besides just integrating logs (with something like Datadog), what's the recommended way of doing application performance monitoring when it comes to deploying node apps on vercel (zeit now v2)?

I know `dd-trace` works for regular deployments, but I'm not sure if that would work in a lambda environment. If we were deploying directly to AWS we could use AWS X-Ray, but since we don't have access to that layer while using vercel, what's the best approach to integrating APM?

Thanks!

**Comment by coetry** [comment]
Hi @aecorredor Datadog actually offers quite comprehensive monitoring technology that goes far beyond logs. They have product offerings that focus on [Serverless Functions](https://www.datadoghq.com/serverless/) and [APM](https://www.datadoghq.com/apm/) in general.

You can really use any service provider for this and all it would take is embedding some tracing coding throughout your Node functions. 

We don't currently have integrations for everything, but welcome community contributions for anything you feel personally inclined to explore :slightly_smiling_face:

**Reply by aecorredor** [reply]
@coetry thanks for the response! Yeap, DD serverless tracing seems to be the correct option. I think my original issue was that I tried to integrate the regular `dd-trace` (meant for long running servers) offering instead of their serverless offering. I'll give that integration a try and report back here once I know if it worked on vercel.

**Comment by aecorredor** [comment]
@coetry so from what I'm reading here https://docs.datadoghq.com/integrations/amazon_web_services/?tab=allpermissions and here https://docs.datadoghq.com/integrations/amazon_lambda, there are some things needed in order for Datadog Serverless to work. We need access to change AWS Policies (which we do not have access to when using vercel), and we also need to set up a Datadog forwarder in the AWS account (which again is not an option in vercel).

I'm guessing some of vercel's enterprise clients must have run into this and they probably have needed Datadog integration. Do you have any insights into what they have done to solve this since accessing AWS directly from vercel is not really an option?

**Comment by aecorredor** [comment]
Ok so after doing more digging and reading the Datadog Lambda docs more carefully, I think this might work: https://docs.datadoghq.com/integrations/amazon_lambda/?tab=nodejs#tracing-with-datadog-apm. They basically use `dd-trace` along with the lambda layer npm package they provide. From looking at the source code it supports the lambda environment by default: https://github.com/DataDog/dd-trace-js/pull/682/files#diff-f3377689fae59b1bfa08b8a4cb493495. It seems like there's no Datadog agent needed for this to work, as they use the log exporter when running on a lambda environment.

**Comment by aecorredor** [comment]
FYI: https://github.com/DataDog/dd-trace-js/issues/934 this is an issue requesting vercel integration support in the `dd-trace` repo. They also mentioned their forwarder is open source and could be used to create an integration.

**Comment by rochdev** [comment]
A bit of background from the Datadog side:

There are currently 3 ways of operation for tracing today:

1) Sending traces to the Datadog agent (the default). This works well for classic environments where the agent can run on the same machine or the same network as the services. While it can be used with serverless as well, it's usually not recommended since the function would be kept running while sending traces, which could increase costs and overall latency.

2) Sending traces to the Datadog API. This is currently only used for browser tracing in beta and would have the same limitations as using the agent, but the latency would be increased even more.

3) Outputting traces to logs, and then use a forwarder to convert the log entries to traces and stream that to Datadog. This is the recommended approach for serverless environments. We only have a [forwarder](https://github.com/DataDog/datadog-serverless-functions/tree/master/aws/logs_monitoring) for AWS at the moment, but it should be straightforward to write a custom forwarder for any source, as long as it's possible to get the logs somehow.

While the best option is definitely `3`, it would require a way to grab the logs from Vercel so that they can be converted to traces and sent to Datadog. Usually this is available as a stream. For example, in AWS it's possible to stream any log using CloudWatch Log Stream to any destination. If something equivalent doesn't exist in Vercel, then the only option right now is most likely using the agent as described in `1` until an integration is developed.

Hope this helps :)
